### PR TITLE
Cluster's destinations are updated by IClusterDestinationsUpdater

### LINKED ISSUE #83: Panic threshold
https://github.com/microsoft/reverse-proxy/pull/78#discussion_r412653804
@davidni: Something to keep in mind is what to do when a large portion of endpoints in a backend are unhealthy. Envoy for example has a `panic mode` where it will prefer to try to route to endpoints it knows to be unhealthy, because it may be better than the alternative of not routing at all. See e.g. [Panic threshold](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/panic_threshold)

### PR DESCRIPTION
**Pull request by alnikola**
Cluster's destination update logic is extracted into a separate service `IClusterDestinationsUpdater` which is now responsible for updating `ClusterDynamicState`, specifically updating the full destination collection and filtering the destinations available for proxying requests to.
This PR also adds `IAvailableDestinationsPolicy` that decides which destinations from the full collection should be available for requests. There are 2 such policies are implemented:
- `StrictHealtyOrUknownDesitnationsPolicy` decides that a destination is available if active AND passive health states are either 'Healthy' or 'Unknown'
- `FallbackToAllDestinationsPolicy` calls `StrictHealtyOrUknownDesitnationsPolicy` as the first step and then checks if the returned available destination collection is empty. If it's **not** empty, this collection is returned as the result, otherwise the full collection containing all destinations is returned. This implements 'panic' health evaluation strategy.

Fixes #83

### CODE REVIEW

File: src/ReverseProxy/Service/HealthChecks/IClusterDestinationsUpdater.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    /// Updates the cluster's destination collections.
+    /// </summary>
+    public interface IClusterDestinationsUpdater
+    {
+        /// <summary>
+        /// Updates the cluster's collection of destination available for proxying requests to.
+        /// Call this if health state has changed for any destinations.
+        /// </summary>
+        /// <param name="cluster">The <see cref="ClusterState"/> owing the destinations.</param>
+        public void UpdateAvailableDestinations(ClusterState cluster);
```
Comment (Kahbazi) [inline src/ReverseProxy/Service/HealthChecks/IClusterDestinationsUpdater.cs:18]:
```suggestion
        void UpdateAvailableDestinations(ClusterState cluster);
```

File: src/ReverseProxy/Service/HealthChecks/IClusterDestinationsUpdater.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+        /// <summary>
+        /// Updates the cluster's collection of all configured destinations.
+        /// Call this after destinations have been added, removed, or their configuration changed.
+        /// This does not need to be called for state updates like health, use UpdateAvailableDestinations for state updates.
+        /// </summary>
+        /// <param name="cluster">The <see cref="ClusterState"/> owing the destinations.</param>
+        public void UpdateAllDestinations(ClusterState cluster);
```
Comment (Kahbazi) [inline src/ReverseProxy/Service/HealthChecks/IClusterDestinationsUpdater.cs:26]:
```suggestion
        void UpdateAllDestinations(ClusterState cluster);
```

File: src/ReverseProxy/Service/HealthChecks/FallbackToAllDestinationsPolicy.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using Yarp.ReverseProxy.Abstractions;
+using Yarp.ReverseProxy.RuntimeModel;
+
+namespace Yarp.ReverseProxy.Service.HealthChecks
+{
+    internal class FallbackToAllDestinationsPolicy : StrictHealthyAndUnknownDestinationsPolicy
```
Comment (Kahbazi) [inline src/ReverseProxy/Service/HealthChecks/FallbackToAllDestinationsPolicy.cs:10]:
```suggestion
    internal sealed class FallbackToAllDestinationsPolicy : StrictHealthyAndUnknownDestinationsPolicy
```

File: src/ReverseProxy/Configuration/DependencyInjection/BuilderExtensions/IReverseProxyBuilderExtensions.cs [hunk 4]
```diff
@@ -38,6 +38,12 @@ public static IReverseProxyBuilder AddRuntimeStateManagers(this IReverseProxyBui
         {
             builder.Services.TryAddSingleton<ITimerFactory, TimerFactory>();
             builder.Services.TryAddSingleton<IDestinationHealthUpdater, DestinationHealthUpdater>();
+
+            builder.Services.TryAddSingleton<IClusterDestinationsUpdater, ClusterDestinationsUpdater>();
+            builder.Services.TryAddEnumerable(new[] {
+                new ServiceDescriptor(typeof(IAvaliableDestinationsPolicy), typeof(StrictHealthyAndUnknownDestinationsPolicy), ServiceLifetime.Singleton),
+                new ServiceDescriptor(typeof(IAvaliableDestinationsPolicy), typeof(FallbackToAllDestinationsPolicy), ServiceLifetime.Singleton)
```
Comment (Kahbazi) [inline src/ReverseProxy/Configuration/DependencyInjection/BuilderExtensions/IReverseProxyBuilderExtensions.cs:45]:
```suggestion
                ServiceDescriptor.Singleton<IAvaliableDestinationsPolicy, StrictHealthyAndUnknownDestinationsPolicy>(),
                ServiceDescriptor.Singleton<IAvaliableDestinationsPolicy, FallbackToAllDestinationsPolicy>()
```

File: src/ReverseProxy/Service/HealthChecks/ClusterDestinationsUpdater.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Yarp.ReverseProxy.RuntimeModel;
+using Yarp.ReverseProxy.Utilities;
+
+namespace Yarp.ReverseProxy.Service.HealthChecks
+{
+    internal class ClusterDestinationsUpdater : IClusterDestinationsUpdater
```
Comment (Kahbazi) [inline src/ReverseProxy/Service/HealthChecks/ClusterDestinationsUpdater.cs:13]:
```suggestion
    internal sealed class ClusterDestinationsUpdater : IClusterDestinationsUpdater
```

**Comment by karelz** [comment]
API review:
- `ClusterDynamicState` -> `ClusterDestinationsState`
    - `HealthyDestinations` -> `AvailableDestinations`
- `ClusterState.DynamicState` -> `ClusterState.DestinationsState`

**Review by Tratcher — COMMENTED** [review]
Will this be added to the health docs?

**Comment by alnikola** [comment]
>Will this be added to the health docs?

Yes. I'm now writing the documentation

File: test/ReverseProxy.Tests/Configuration/ConfigurationConfigProviderTests.cs [hunk 6]
```diff
@@ -216,7 +216,8 @@ public class ConfigurationConfigProviderTests
                     ""Timeout"": ""00:00:06"",
                     ""Policy"": ""Any5xxResponse"",
                     ""Path"": ""healthCheckPath""
-                }
+                },
+                ""AvailableDestinationsPolicy"": ""FallbackToAll""
```
Comment (Tratcher) [inline test/ReverseProxy.Tests/Configuration/ConfigurationConfigProviderTests.cs:220]:
```suggestion
                ""AvailableDestinationsPolicy"": ""HealthyOrPanic""
```

File: test/ReverseProxy.Tests/Service/RuntimeModel/ClusterStateTests.cs [hunk 7]
```diff
@@ -2,18 +2,15 @@
...
-using System.Threading;
-using System.Threading.Tasks;
 using Moq;
 using Xunit;
 using Yarp.ReverseProxy.Abstractions;
 
 namespace Yarp.ReverseProxy.RuntimeModel.Tests
 {
-    public class ClusterInfoTests
+    /*public class ClusterStateTests
```
Comment (Tratcher) [inline test/ReverseProxy.Tests/Service/RuntimeModel/ClusterStateTests.cs:13]: Delete, this is all covered elsewhere now.

File: src/ReverseProxy/Service/HealthChecks/DestinationHealthUpdater.cs [hunk 8]
```diff
@@ -64,7 +69,13 @@ internal Task SetPassiveAsync(ClusterState cluster, DestinationState destination
             {
                 healthState.Passive = newHealth;
                 ScheduleReactivation(cluster, destination, newHealth, reactivationPeriod);
-                return Task.Factory.StartNew(c => ((ClusterState)c!).UpdateDynamicState(), cluster, TaskCreationOptions.RunContinuationsAsynchronously);
+                return Task.Factory.StartNew(
+                    i => {
+                        var k = ((ClusterState Cluster, IClusterDestinationsUpdater Updater))i!;
+                        k.Updater.UpdateAvailableDestinations(cluster);
+                    },
```
Comment (Tratcher) [inline src/ReverseProxy/Service/HealthChecks/DestinationHealthUpdater.cs:76]: _clusterUpdater is a class variable so this delegate could be replaced by a class method `private void UpdateCluster(object obj)` that only passes in the cluster.

File: src/ReverseProxy/Service/HealthChecks/HealthCheckConstants.cs [hunk 9]
```diff
@@ -14,5 +14,22 @@ public static class ActivePolicy
...
+            /// <summary>
+            /// Marks destination as available for proxying requests to if its health state
+            /// is either 'Healthy' or 'Unknown'.
+            /// </summary>
+            /// <remarks>It applies only if active or passive health checks are enabled.</remarks>
+            public static readonly string HealthyAndUnknown = nameof(HealthyAndUnknown);
+
+            /// <summary>
+            /// Calls <see cref="HealthyAndUnknown"/> policy at first to determine
+            /// desninations' availability. If no available destinations are returned
```
Comment (Tratcher) [inline src/ReverseProxy/Service/HealthChecks/HealthCheckConstants.cs:29]:
```suggestion
            /// destinations' availability. If no available destinations are returned
```

File: src/ReverseProxy/Service/HealthChecks/HealthCheckConstants.cs [hunk 10]
```diff
@@ -14,5 +14,22 @@ public static class ActivePolicy
         {
             public static readonly string ConsecutiveFailures = nameof(ConsecutiveFailures);
         }
+
+        public static class AvailableDestinations
+        {
+            /// <summary>
+            /// Marks destination as available for proxying requests to if its health state
+            /// is either 'Healthy' or 'Unknown'.
```
Comment (Tratcher) [inline src/ReverseProxy/Service/HealthChecks/HealthCheckConstants.cs:22]:
```suggestion
            /// is either 'Healthy' or 'Unknown'. If no destinations are available then requests will get a 503 error.
```

File: src/ReverseProxy/Service/HealthChecks/HealthyAndUnknownDestinationsPolicy.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Yarp.ReverseProxy.Abstractions;
+using Yarp.ReverseProxy.RuntimeModel;
+
+namespace Yarp.ReverseProxy.Service.HealthChecks
+{
+    // Policy marking destinations as available only if their active and passive health states
+    /// are either 'Healthy' or 'Unknown'/>.
+    internal class StrictHealthyAndUnknownDestinationsPolicy : IAvailableDestinationsPolicy
```
Comment (Tratcher) [inline src/ReverseProxy/Service/HealthChecks/HealthyAndUnknownDestinationsPolicy.cs:13]:
```suggestion
    internal class HealthyAndUnknownDestinationsPolicy : IAvailableDestinationsPolicy
```

Comment (Tratcher) [reply src/ReverseProxy/Service/HealthChecks/HealthyAndUnknownDestinationsPolicy.cs:13]: Are we going to want this to be public as the base class for other custom policies like we did for HealthyOrPanicDestinationsPolicy ?

File: src/ReverseProxy/Service/HealthChecks/HealthyOrPanicDestinationsPolicy.cs [hunk 12]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using Yarp.ReverseProxy.Abstractions;
+using Yarp.ReverseProxy.RuntimeModel;
+
+namespace Yarp.ReverseProxy.Service.HealthChecks
+{
+    internal sealed class FallbackToAllDestinationsPolicy : StrictHealthyAndUnknownDestinationsPolicy
```
Comment (Tratcher) [inline src/ReverseProxy/Service/HealthChecks/HealthyOrPanicDestinationsPolicy.cs:10]:
```suggestion
    internal sealed class HealthyOrPanicDestinationsPolicy : HealthyAndUnknownDestinationsPolicy
```

File: src/ReverseProxy/Service/HealthChecks/HealthyAndUnknownDestinationsPolicy.cs (hunk 11 above)
Comment (alnikola) [reply src/ReverseProxy/Service/HealthChecks/HealthyAndUnknownDestinationsPolicy.cs:13]: I don't think it needs to be public right now. I'd like to wait for an actual customer request before exposing that. As a note, we also didn't expose BaseSessionAffinityProvider which serves a similar purpose.

File: src/ReverseProxy/Abstractions/HealthCheckConfig.cs [hunk 13]
```diff
@@ -29,13 +34,14 @@ public bool Equals(HealthCheckConfig? other)
...
-                && Active == other.Active;
+                && Active == other.Active
+                && string.Equals(AvailableDestinationsPolicy, other.AvailableDestinationsPolicy, StringComparison.OrdinalIgnoreCase);
         }
 
         /// <inheritdoc />
         public override int GetHashCode()
         {
-            return HashCode.Combine(Passive, Active);
+            return HashCode.Combine(Passive, Active, AvailableDestinationsPolicy?.GetHashCode());
```
Comment (davidfowl) [inline src/ReverseProxy/Abstractions/HealthCheckConfig.cs:44]:
```suggestion
            return HashCode.Combine(Passive, Active, AvailableDestinationsPolicy);
```

This handles null.

**Comment by alnikola** [comment]
PR is ready for review.

**Review by Tratcher — APPROVED** [review]

**Comment by davidfowl** [comment]
Have we run any performance tests to make sure there's no obvious problems when this is on?

**Comment by Tratcher** [comment]
Not yet. This all intentionally runs on background threads to avoid blocking requests. Alexander is creating a work item to profile the passive health check logic and check for per destination bottlenecks when calculating stats before jumping to the background.

**Comment by davidfowl** [comment]
It's still allocating tasks (and maybe other things). I'd like us to just make sure there's nothing obviously bad that jumps out when these are on.

File: docs/docfx/articles/dests-health-checks.md [hunk 14]
```diff
@@ -1,5 +1,5 @@
...
+In most of the real-world systems, it's expected for their nodes to occasionally experience transient issues and go down completely due to a variety of reasons such as an overload, resource leakage, hardware failures, etc. Ideally, it'd be desirable to completely prevent those unfortunate events from occurring in a proactive way, but the cost of designing and building such an ideal system is generally prohibitively high. However, there is another reactive approach which is cheaper and aimed to minimizing a negative impact failures cause on client requests. The proxy can analyze each nodes health and stop sending client traffic to unhealthy ones until they recover. YARP implements this approach in the form of active and passive destination health checks. They are indpendent from each other and stored on the relative properties for each destinations. Health states are initialized with `Unkown` value which can be later changed to either `Healty` or `Unhealthy` by the corresponding polices as explained below.
```
Comment (MihaZupan) [inline docs/docfx/articles/dests-health-checks.md:2]:
```suggestion
In most of the real-world systems, it's expected for their nodes to occasionally experience transient issues and go down completely due to a variety of reasons such as an overload, resource leakage, hardware failures, etc. Ideally, it'd be desirable to completely prevent those unfortunate events from occurring in a proactive way, but the cost of designing and building such an ideal system is generally prohibitively high. However, there is another reactive approach which is cheaper and aimed to minimizing a negative impact failures cause on client requests. The proxy can analyze each nodes health and stop sending client traffic to unhealthy ones until they recover. YARP implements this approach in the form of active and passive destination health checks. They are independent from each other and stored on the relative properties for each destination. Health states are initialized with `Unknown` value which can be later changed to either `Healty` or `Unhealthy` by the corresponding policies as explained below.
```

File: docs/docfx/articles/dests-health-checks.md [hunk 15]
```diff
@@ -321,10 +321,58 @@ public class FirstUnsuccessfulResponseHealthPolicy : IPassiveHealthCheckPolicy
...
+## Available destination collection
+Destinations health state is used to determine which of them are eligible for receiving proxied requests. Each cluster maintains it's own list of available destinations on `AvailableDestinations` property of [ClusterDestinationState](xref:Yarp.ReverseProxy.RuntimeModel.ClusterDestinationsState) type. That list gets rebuilt when any destination's health state changes. The [IClusterDestinationsUpdater](xref:Yarp.ReverseProxy.Service.HealthChecks.IClusterDestinationsUpdater) controls that process and calls a [IAvailableDestinationsPolicy](xref:Yarp.ReverseProxy.Service.HealthChecks.IAvailableDestinationsPolicy) configured on the cluster to actually choose the available destinations from the all cluster's destinations. There are the following built-in policies provided and custom ones can be implemented if necessary.
```
Comment (MihaZupan) [inline docs/docfx/articles/dests-health-checks.md:325]:
```suggestion
Destinations health state is used to determine which of them are eligible for receiving proxied requests. Each cluster maintains its own list of available destinations on `AvailableDestinations` property of the [`ClusterDestinationState`](xref:Yarp.ReverseProxy.RuntimeModel.ClusterDestinationsState) type. That list gets rebuilt when any destination's health state changes. The [`IClusterDestinationsUpdater`](xref:Yarp.ReverseProxy.Service.HealthChecks.IClusterDestinationsUpdater) controls that process and calls an [`IAvailableDestinationsPolicy`](xref:Yarp.ReverseProxy.Service.HealthChecks.IAvailableDestinationsPolicy) configured on the cluster to actually choose the available destinations from the all cluster's destinations. There are the following built-in policies provided and custom ones can be implemented if necessary.
```

File: docs/docfx/articles/dests-health-checks.md [hunk 16]
```diff
@@ -321,10 +321,58 @@ public class FirstUnsuccessfulResponseHealthPolicy : IPassiveHealthCheckPolicy
...
 
-- Active and passive health state are initialized with `Unkown` value which can be later changed to either `Healty` or `Unhealthy` by the corresponding polices.
-- A destination is added on the healthy destination list when all of the following is TRUE:
-  - Active health checks are disabled on the cluster OR `DestinationHealthState.Active != DestinationHealth.Unhealthy`
-  - Passive health checks are disabled on the cluster OR `DestinationHealthState.Passive != DestinationHealth.Unhealthy`
+- `HealthyAndUnknown` - inspects each `DestinationState` and adds it on the available destination list if all of the following statements is TRUE. If no destinations are available then requests will get a 503 error. This is the default policy.
```
Comment (MihaZupan) [inline docs/docfx/articles/dests-health-checks.md:327]:
```suggestion
- `HealthyAndUnknown` - inspects each `DestinationState` and adds it on the available destination list if all of the following statements are TRUE. If no destinations are available then requests will get a 503 error. This is the default policy.
```

File: docs/docfx/articles/dests-health-checks.md [hunk 17]
```diff
@@ -321,10 +321,58 @@ public class FirstUnsuccessfulResponseHealthPolicy : IPassiveHealthCheckPolicy
...
+#### Code example
+```C#
+var clusters = new[]
+{
+    new ClusterConfig()
+    {
+        ClusterId = "cluster1",
+        HealthCheck = new HealthCheckConfig
+        {
+            AvailableDestinationsPolicy = HealthCheckConstants.AvailableDestinations.HealthyOrPanic;
```
Comment (MihaZupan) [inline docs/docfx/articles/dests-health-checks.md:365]:
```suggestion
            AvailableDestinationsPolicy = HealthCheckConstants.AvailableDestinations.HealthyOrPanic,
```

**Review by MihaZupan — APPROVED** [review]

**Comment by alnikola** [comment]
>Alexander is creating a work item to profile the passive health check logic and check for per destination bottlenecks when calculating stats before jumping to the background.

Submitted the issue #1033

**Pull request merged**